File: src/main/java/org/trp/shincolle/entity/base/ShipGuardTarget.java
package org.trp.shincolle.entity.base;

import net.minecraft.core.BlockPos;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.Vec3;

public record ShipGuardTarget(int x, int y, int z, int dimensionId, Type type) {
    public static final ShipGuardTarget NONE = new ShipGuardTarget(-1, -1, -1, 0, Type.NONE);

    public static ShipGuardTarget fromShip(EntityShipBase ship) {
        return new ShipGuardTarget(
                ship.getGuardedPos(0),
                ship.getGuardedPos(1),
                ship.getGuardedPos(2),
                ship.getGuardedPos(3),
                Type.fromLegacy(ship.getGuardedPos(4))
        );
    }

    public boolean isActive() {
        return this.type != Type.NONE;
    }

    public boolean isBlock() {
        return this.type == Type.BLOCK;
    }

    public boolean isEntity() {
        return this.type == Type.ENTITY;
    }

    public boolean isIn(Level level) {
        return this.dimensionId == EntityShipBase.getLegacyDimensionId(level);
    }

    public BlockPos blockPos() {
        return new BlockPos(this.x, this.y, this.z);
    }

    public Vec3 blockCenter() {
        return new Vec3(this.x + 0.5D, this.y, this.z + 0.5D);
    }

    public int legacyType() {
        return this.type.legacyId;
    }

    public enum Type {
        NONE(0),
        BLOCK(1),
        ENTITY(2);

        private final int legacyId;

        Type(int legacyId) {
            this.legacyId = legacyId;
        }

        public int legacyId() {
            return this.legacyId;
        }

        static Type fromLegacy(int legacyId) {
            return switch (legacyId) {
                case 1 -> BLOCK;
                case 2 -> ENTITY;
                default -> NONE;
            };
        }
    }
}

File: src/main/java/org/trp/shincolle/entity/base/ShipGuardTarget.kt
package org.trp.shincolle.entity.base

import net.minecraft.core.BlockPos
import net.minecraft.world.level.Level
import net.minecraft.world.phys.Vec3

@JvmRecord
data class ShipGuardTarget(val x: Int, val y: Int, val z: Int, val dimensionId: Int, val type: Type?) {
    val isActive: Boolean
        get() = this.type != Type.NONE

    val isBlock: Boolean
        get() = this.type == Type.BLOCK

    val isEntity: Boolean
        get() = this.type == Type.ENTITY

    fun isIn(level: Level): Boolean {
        return this.dimensionId == EntityShipBase.Companion.getLegacyDimensionId(level)
    }

    fun blockPos(): BlockPos {
        return BlockPos(this.x, this.y, this.z)
    }

    fun blockCenter(): Vec3 {
        return Vec3(this.x + 0.5, this.y.toDouble(), this.z + 0.5)
    }

    fun legacyType(): Int {
        return this.type.legacyId
    }

    enum class Type(private val legacyId: Int) {
        NONE(0),
        BLOCK(1),
        ENTITY(2);

        fun legacyId(): Int {
            return this.legacyId
        }

        companion object {
            fun fromLegacy(legacyId: Int): Type {
                return when (legacyId) {
                    1 -> Type.BLOCK
                    2 -> Type.ENTITY
                    else -> Type.NONE
                }
            }
        }
    }

    companion object {
        val NONE: ShipGuardTarget = ShipGuardTarget(-1, -1, -1, 0, Type.NONE)

        fun fromShip(ship: EntityShipBase): ShipGuardTarget {
            return ShipGuardTarget(
                ship.getGuardedPos(0),
                ship.getGuardedPos(1),
                ship.getGuardedPos(2),
                ship.getGuardedPos(3),
                Type.Companion.fromLegacy(ship.getGuardedPos(4))
            )
        }
    }
}
